jaraco/windows/clipboard.py

```diff
@@ -82,42 +80,73 @@ class HTMLSnippet(object):
+    END_FRAGMENT_MARKER = b'<!--EndFragment-->'
+
     def __init__(self, handle):
-        self.data = nts(raw_data(handle).decode('utf-8'))
+        self.data = raw_data(handle)
         self.headers = self.parse_headers(self.data)
 
     @property
     def html(self):
-        return self.data[self.headers['StartHTML'] :]
+        start_html = self.headers['StartHTML']
+        end_html = self.headers['EndHTML']
+        if start_html == -1:
+            start_html = self.headers['StartFragment'] - len(
+                HTMLSnippet.START_FRAGMENT_MARKER
+            )
+        if end_html == -1:
+            end_html = self.headers['EndFragment'] + len(
+                HTMLSnippet.END_FRAGMENT_MARKER
+            )
+        return self.data[start_html:end_html].decode('utf-8')
 
     @property
     def fragment(self):
-        return self.data[self.headers['StartFragment'] : self.headers['EndFragment']]
+        start_fragment = self.headers['StartFragment']
+        end_fragment = self.headers['EndFragment']
+        return self.data[start_fragment:end_fragment].decode('utf-8')
+
+    @property
+    def selection(self):
+        start_selection = (
+            self.headers.get('StartSelection') or self.headers['StartFragment']
+        )
+        end_selection = self.headers.get('EndSelection') or self.headers['EndFragment']
+        return self.data[start_selection:end_selection].decode('utf-8')
 
     @staticmethod
     def parse_headers(data):
-        d = io.StringIO(data)
-
-        def header_line(line):
-            return re.match(r'(\w+):(.*)', line)
-
-        headers = map(header_line, d)
-        # grab headers until they no longer match
-        headers = itertools.takewhile(bool, headers)
-
         def best_type(value):
             try:
                 return int(value)
             except ValueError:
-                pass
-            try:
-                return float(value)
-            except ValueError:
-                pass
-            return value
-
-        pairs = ((header.group(1), best_type(header.group(2))) for header in headers)
-        return dict(pairs)
+                return value.decode('utf-8')
+
+        match = re.match(rb'Version:(?:0\.9|1\.0)(?:\r\n|\r|\n)', data)
+        if not match:
+            raise ValueError('Unsupported format')
+        pattern = re.compile(rb'(\w+):([^\r\n]*)(?:\r\n|\r|\n)')
+        pos = match.end()
+        endpos = len(data)
+        match = pattern.match(data, pos, endpos)
+        headers = {}
+        while match:
+            key = match.group(1).decode('utf-8')
+            value = best_type(match.group(2))
+            headers[key] = value
+            pos = match.end()
+            if key == 'StartHTML' and value != -1:
+                endpos = value
+            elif key == 'StartFragment':
+                endpos = min(endpos, value - len(HTMLSnippet.START_FRAGMENT_MARKER))
+            match = pattern.match(data, pos, endpos)
+        return headers
```

Comment:
I try to avoid the init/for/append pattern, which is why the previous code used iterators and a dict constructor to construct the dict in one expression. I'd like not to lose that pattern.
I realize that pattern isn't something that a lot of people are comfortable with, so I'll plan to refactor this to my preference, unless you'd like to take a stab at it (let me know).